### ISSUE TITLE
[Feature request] loop (shuffle) button

### ISSUE THREAD

**Issue by Solarunit** [DeaDBeeF-Player/deadbeef#1992]
Please add loop (shuffle) button. It would be nice to have some indication on the main window (or maybe tray area) when loop mode is enabled.

**Comment by sklprogs** [comment]
As to me, it would be great to have "loop all/track" option in the program tray (since I have the main window closed most of the time). I use 0.7.2, and the "shuffle" option, unlike "loop" option, is already there. Quite frustrating, since I frequently use "loop", but never - "shuffle".

**Comment by fabertawe** [comment]
Partly a "me too" to having indication of shuffle/loop status on the status bar but also to suggest possibly exposing the current status to a variable (title formatting)?

I've made custom buttons for the various shuffle/loop options but have no way to know the current status of shuffle/loop. I could use a variable in the output of the "playback status" plugin.

**Comment by kpcee** [comment]
There is a plugin called ["Playback order menu"](https://deadbeef.sourceforge.io/plugins.html), it would be easy to extend it or make a new one to show the repeat status.

**Comment by fabertawe** [comment]
@kpcee - thank you so much! I don't know how I missed that plugin.

You say "easy" ;) ...I'll take a look and see if I can duplicate it for repeat status.

**Comment by kpcee** [comment]
In the meantime I have created a corresponding plugin myself. It creates two buttons with playback order and playback loop, if you click on them you can change the mode to Loop track → Loop all and Linear → Shuffle. The plugin runs with Gtk 2 without problems, but under Gtk 3 it crashes after a few clicks. This is because the click signal of the button triggers several signals where the label of the button is changed and DeaDBeeF doesn't like that at all.
In later versions it could also be possible to replace the text with a graphic, so that it looks like Audacious. 
Does anyone have a suggestion for a name for the plugin?
![pic](https://user-images.githubusercontent.com/39984895/71825471-3e829c80-309c-11ea-956f-9a133b48a258.png)

**Comment by fabertawe** [comment]
That looks great, I'd love to test it. I only use Gtk2 myself (where possible).

As for a name... "Shoop" ;) What about something along the lines of "(playback) control button(s)"? Just pick something you're happy with.

**Comment by kpcee** [comment]
Okay I just created a new repository [deadbeef-playback-buttons](https://github.com/kpcee/deadbeef-playback-buttons).

**Comment by sklprogs** [comment]
> As for a name... "Shoop" ;)
This is a slang word for "to have sex". Shloop/loosh?

**Comment by fabertawe** [comment]
Apologies for the delay in getting back on this.

It works great, many thanks!

The only thing your missing is random album and random track from the shuffle button itself but the button does display the correct status when these are selected elsewhere (menu). Similarly, with the loop button, it doesn't go to "Loop Off" unless selected from the menu.

Replacing the text (as you alluded to) at some point with a graphic would be ideal. I'm not keen on "Linear" as I'd prefer "Off". Also, could the button just display "Album" instead of "RND Album", for consistency? Minor niggles, that's all :)

Great work, this has filled a gap for me. Deadbeef is almost perfect now :)

**Comment by kpcee** [comment]
Glad you like it :)
The fact that you can't reach all the play modes from the buttons is due to personal preferences, I don't use album and random, on the other hand a graphical button would only switch between Off and Shuffle or Repeat Track / All. But it's not hard to change that if it's too annoying ;)
I will change the names of the buttons in the next version. Theoretically, I could also add new play modes, like Keep Artist, Keep Album, Top Rated Songs et al.

The difference between Random and Shuffle is well explained in the [foobar2000-FAQ](https://www.foobar2000.org/FAQ#random_shuffle_playback_order).

**Comment by fabertawe** [comment]
> I will change the names of the buttons in the next version. Theoretically, I could also add new play modes, like Keep Artist, Keep Album, Top Rated Songs et al.

"Keep Artist" sounds interesting. I don't rate songs but other's do and this could be useful to a lot of people. It does what I need already so anything else is a bonus :)

> The difference between Random and Shuffle is well explained in the [foobar2000-FAQ](https://www.foobar2000.org/FAQ#random_shuffle_playback_order).

Ah, it makes sense now, thanks.

**Comment by Alexey-Yakovenko** [comment]
@kpcee deadbeef's shuffle works slightly differently though.
Foobar FAQ claims that in shuffle mode, the playlist would be shuffled once, and then play in that order repeatedly.
In deadbeef, after playing the shuffled playlist, it will get re-shuffled again.

There are possibly other differences too.

**Comment by kpcee** [comment]
> There are possibly other differences too.

Yes, e.g. songs that have already been played are ignored in the search, so the order is never identical if you always start the same track manually.

**Comment by erdnuesse** [comment]
Any news on implementing it in master branch?

**Comment by kpcee** [comment]
I have uploaded a new version of my [plugin](https://github.com/kpcee/deadbeef-playback-buttons) which has now three buttons, the new button adds new playback modes ("Keep Album", "Keep Artist", "Top Rated", "Selection", "Playlist" deactivates the plugin). It works internally with its own playlist and supports the suffle modes "Linear", "Shuffle" and "Random".
There is also a new entry in the context menu called "Custom Playlist".
![Unbenannt](https://user-images.githubusercontent.com/39984895/94568965-773b5500-026d-11eb-9aee-2e884aa9db20.png)

**Comment by Storme9** [comment]
My problem is that I am a noob.

I have DeadBeef installed but I am unable to follow the instructions in the ReadMe, specifically:

"To compile the plugin you need to copy the files deadbeef.h and gtkui_api.h from the deadbeef directory." - I don't think that copying files compiles anything. What am I supposed to do with the copied files? Copy them to where?

and secondly:

"Copy the compiled plugin to the plugin folder (~/.local/lib/deadbeef/) and restart DeadDBeeF, then add the plugin to the gui." - to this I ask:

    How do I "compile" the plugin?
    How do I add the plugin to the gui?

Any direction would be much appreciated since I want the feature of your plugin and I also want to learn about my operating system.

Deadbeef version: 1.8.8
OS: Ubuntu 20.04

**Comment by Alexey-Yakovenko** [comment]
@Storme9 this particular plugin requires a copy of some files from deadbeef source, in order to compile.
You need to copy the files from deadbeef source folder into the plugin source folder.
After compiling and installing the plugin, assuming it works, you can add it to the UI using "Design Mode" -- which is accessible from the View menu. You will need to customize the UI, and rearrange some UI elements, to add this new UI element where you like it.

**Comment by kpcee** [comment]
@Storme9 Maybe I should also provide compiled versions. Did everything work now?

**Comment by Storme9** [comment]
As soon as I get a moment - I will try and do what you said and then let
you know how it worked out.

**Comment by Storme9** [comment]
So I am trying to figure out how to install your plugin.

I started by trying to locate the source files of my deadbeef install. I
did this with the following command: whereis deadbeef

It came up with 3 locations: /usr/bin/deadbeef /usr/lib/deadbeef
/usr/share/deadbeef

Note: deadbeef is not installed in the ~/.local/lib/deadbeef/ location as
you suggest.

listing all items in these directories resulted in zero files with the
names: deadbeef.h and gtkui_api.h

I am dead in the water...

Attached is a screenshot of the items in the directory: /usr/bin/deadbeef

I would really appreciate some help please.

**Comment by Alexey-Yakovenko** [comment]
You can get the files you need from this archive:
https://sourceforge.net/projects/deadbeef/files/staticdeps/ddb-headers-latest.tar.bz2/download

**Comment by Storme9** [comment]
So I have those files now and I have copied them to the folder containing
your plugin files. (please see screenshot)

Now I need to "compile" the plugin. I have never done this before. My plan
is to follow this guide:
https://www.howtogeek.com/105413/how-to-compile-and-install-from-source-on-ubuntu/
up until the end of the "compile" step. Is this okay?

And then finally your instructions say: "Copy the compiled plugin to the
plugin folder (`~/.local/lib/deadbeef/`) and restart DeadDBeeF, then add
the plugin to the gui." Where is the plugin folder?

**Comment by Alexey-Yakovenko** [comment]
for compiling the plugin, typically it's enough to just run `make`

to copy: 
```
mkdir -p ~/.local/lib/deadbeef/
cp "plugin_name.so" ~/.local/lib/deadbeef/
```

(change the .so file name to the name of compiled plugin)

**Comment by Storme9** [comment]
I'm assuming I need to install gtk2.0 - is this correct?

**Comment by Storme9** [comment]
I seem to already have gtk2.0... Please see screenshot.

**Comment by Alexey-Yakovenko** [comment]
There's no screenshot in your message.
And yes, for this plugin you need a "dev" package of GTK to be installed: `sudo apt install libgtk-3-dev`

**Comment by Storme9** [comment]
compilation complete.

I copied the gtk2 plugin (it's a .so file) to the directory you told me to
create.

Upon starting deadbeef I got an error even though deadbeef started.

I uninstalled and then reinstalled deadbeef only to be greeted by errors.
(please see attached screenshot.

Upon google searching the errors - I get back to a deadbeef thread: "If
you're sure the reason is deadbeef - make sure to delete all 3rd party
deadbeef plugin, and provide a gdb backtrace. Instruction how to do it can
be found here: https://github.com/Alexey-Yakovenko/deadbeef/wiki/GDB-HOWTO"

that link does not help...

I would appreciate another pointer please.

Would you suggest uninstalling all gtk packages and deadbeef and then
reinstalling?

**Comment by Alexey-Yakovenko** [comment]
there's no screenshot in your message. It seems like email replies discard the attachments

**Comment by Oleksiy-Yakovenko** [comment]
@Storme9 can you please remind what is the problem again / what do you need help with? This conversation happened over 5 months ago, and I lost track of what it was about. I don't see any screenshot in any of your messages, and I'm not sure what problem you're trying to solve.

**Comment by Storme9** [comment]
It seems you have added a repeat menu to the player which solves all issues. Thank you.

**Comment by DracoDynasty** [comment]
Any news on that feature ?

The third party plugin cited in this thread doesn't compile, giving a bunch of `[-Wincompatible-pointer-types]` errors. Possibly because it's outdated since it's been half a decade ?

Besides, if I understand correctly, that plugin makes its own internal playlist to work, which would be disheartening to users who will obviously want it to work with their already created playlists.

I compiled and installed the plugin from `https://github.com/cboxdoerfer/ddb_playback_order` and it solves half the problem: shuffle status is now displayed on the interface, which is a marked improvement and exactly what I wanted for the shuffle !

The part still missing after installing that plugin is the same kind of feature for repeat status (none/one/all).

**Comment by Oleksiy-Yakovenko** [comment]
@DracoDynasty have you tried the other plugin linked from https://github.com/DeaDBeeF-Player/deadbeef/issues/1992#issuecomment-571581226 ?

**Comment by DracoDynasty** [comment]
@Oleksiy-Yakovenko Yes, I have. That is the one I was talking about when I said that it didn't compile and gave a bunch of errors. Not sure what was up with that since I don't have enough knowledge to go and debug it.

**Comment by Oleksiy-Yakovenko** [comment]
@DracoDynasty you should be able to compile the plugin by running this command:
```
CFLAGS="-Wno-error -Wno-incompatible-pointer-types" make
```

**Comment by DracoDynasty** [comment]
@Oleksiy-Yakovenko Thanks for the command, which did successfully let me compile the plugin.

Sadly, a quick test of the plugin revealed a bug that crashes DeaDBeeF when anything happens that changes the `Random` or `Playlist` button while a track is playing.
This happens both when you directly click the buttons a few times, or if you use a keyboard shortcut to cycle through the shuffle playback options.

To me it seems that any action that triggers to quick of a change in certain button labels will crash the program with a segfault, which sadly makes it undesirable as a solution.

**Comment by Oleksiy-Yakovenko** [comment]
Sorry to hear. Unfortunately I can't provide any solution right at this moment, but I'll see if I can find some time to spend on the feature after holidays.

**Comment by DracoDynasty** [comment]
That is absolutely okay. Thank you very much for helping me install and test that plugin.

And I wish you nice holidays.

**Comment by kpcee** [comment]
I have updated the version on GitHub; it should run without any problems now.

**Comment by DracoDynasty** [comment]
@kpcee Thank you for the update !

The bug still happens to me, sadly.
I pulled the latest version from git, did `make clean` then `make gtk3` then copied the library to the locations.
Clicking the shuffle button a few times still crashes DeaDBeeF in my testing. I'm not sure what exactly causes it.

**Comment by kpcee** [comment]
Okay, thanks for trying it out. I can confirm the crash for GTK3. The GTK2 version does not seem to be affected by the bug anymore. Clicking a button several times in quick succession, e.g., Linear/Shuffle, causes the crash.

I can also say what exactly causes the crash, namely changing the text of the button. If you comment out the line
`gtk_button_set_label(GTK_BUTTON(widget), text);`

the text of the button is no longer changed, but the plugin itself continues to run without crashing. I don't know how to fix this. 🤷

**Comment by DracoDynasty** [comment]
That's a very weird bug indeed. Thanks for taking the time to look into it !

If it's only this button, maybe we could remove just this button and use the other plugin ([ddb_playback_order](https://github.com/cboxdoerfer/ddb_playback_order)) in its place ?

If all buttons can cause the bug on text update, then it kind of defeats the purpose of the plugin on my end, which was to get a constant on-interface display of the current loop status since I already have the other plugin for shuffle order.

I am as baffled as you are as to why a simple text change would cause the entire program to segfault...
Only thing I can think of is, if the update is done asynchronously maybe having quick changes somehow ends up in a rare race condition that triggers both updates concurrently in different threads.

If that was it, a fix would be to put some semaphore on text change for each button so that the first call has to finish its update before the next one completes... but then if they somehow end up running in the wrong order, the label of the button could show the wrong state.
In that case, maybe getting the time of the call itself (the moment the button press event is triggered) and saving it for each button would work.
When you want to update the label, you'd first compare the event's trigger time to the one stored in the button. If it's older, you skip the update as to not put on the label data that is older that what it currently has.

Anyway, those are just ideas. I have no idea how your code is structured, if GTK events even store their trigger time, or if a async updates are even the problem in the first place. Just food for thoughts really.

**Comment by Oleksiy-Yakovenko** [comment]
@kpcee this bug occurs because you're expected to update UI on main thread, but playback_buttons_message is not called on main thread.
You need to wrap the ui code in a function, and dispatch it via `g_idle_add`

**Comment by kpcee** [comment]
Okay, thanks for letting me know. 👍
I've updated the plugin.

**Comment by DracoDynasty** [comment]
@kpcee Sorry for taking this long to answer. I had some... unpleasant start of the year. Let's just leave it at that.

Anyway, I'm happy to report that your plugin now works without crashing DeaDBeeF at all !
Many thanks for the quick update :) !

**Comment by kpcee** [comment]
Great to hear that the plugin is now running without crashes - thanks for the feedback! 😊
